triton-vm/src/table/master_table.rs

```diff
@@ -927,10 +898,35 @@ impl MasterBaseTable {
     pub fn extend(&self, challenges: &Challenges) -> MasterExtTable {
         // randomizer polynomials
         let num_rows = self.randomized_trace_table().nrows();
-        let mut randomized_trace_extension_table = Array2::zeros([num_rows, NUM_EXT_COLUMNS].f());
+        profiler!(start "initialize master table");
+        let num_table_elements = num_rows * NUM_EXT_COLUMNS;
+        let mut randomized_trace_extension_table = Vec::with_capacity(num_table_elements);
+
+        #[allow(clippy::uninit_vec)]
+        unsafe {
+            // Use of unsafe code because it is faster than the alternative, which is
+            // `vec![XFieldElemet::zero(); num_table_elements]` and which is slower
+            // because it is fully sequential.
+            // This use of unsafe code is safe because
+            //  - We have enough capacity to change the length (because the variable `num_table_elements`
+            //    is immutable and is not being shadowed.)
+            //  - All elements are initialized in the next few lines.
+            randomized_trace_extension_table.set_len(num_table_elements);
+        }
+        let mut randomized_trace_extension_table = Array2::from_shape_vec(
+            [num_rows, NUM_EXT_COLUMNS].f(),
+            randomized_trace_extension_table,
+        )
+        .unwrap();
+
+        profiler!(start "memory warmup");
+        randomized_trace_extension_table.par_mapv_inplace(|_| XFieldElement::zero());
+        profiler!(stop "memory warmup");
```

Comment:
Is there a benchmark comparing these respective solutions in terms of speed?

Reply:
Flagging @Sword-Smith for comment since he came up with this snippet.

Reply:
Thanks for asking for benchmarks. They are interesting. Time in below table is in μs.

|log₂(size)    |  10     | 12      |14     | 16     | 18     | 20      |
|:-------------|---------:|---------:|-------:|--------:|--------:|---------:|
|default       |  0.22  | 0.8 |3.16 | 16.37 | 65.405 | 351.98  |
|set_len       |  198.79 | 283.11  |434.6  | 602.63 | 826.58 | 1131.7  |
|maybe_uninit  |  199.41 | 284.61  |432.83 | 597.93 | 826.85 | 1124.4  |

![image](https://github.com/TritonVM/triton-vm/assets/4719768/1e95f607-0a91-4798-a53f-63a0e8ffdab6)

Reply:
What's even more concerning than performance is that the current code might not fulfill the safety contract of `.set_len()`. Quoting the `Safety` section from the documentation (emphasis mine):

> - `new_len` must be less than or equal to `capacity()`.
> - The elements at `old_len..new_len` **must be initialized**.

It is not completely clear to me whether this means (a) the elements at `old_len..new_len` must _already_ be initialized, or whether it means that (b) the elements at `old_range..new_range` must be initialized _at some point in the future_. In the example provided in the documentation, the elements are already initialized, suggesting that interpretation (a) is correct.

I do not want to include `unsafe` code where I am uncertain whether the safety contract is being upheld. Unfortunately, this also calls into question the already-present `unsafe` call to `.set_len()`.

Reply:
I added branches to facilitate benchmarking.
 - [asz/parallel-table-extension-option-zeros-f](https://github.com/TritonVM/triton-vm/tree/asz/parallel-table-extension-option-zeros-f)
 - [asz/parallel-table-extension-option-set-len](https://github.com/TritonVM/triton-vm/tree/asz/parallel-table-extension-option-set-len)
 - [asz/parallel-table-extension-option-maybe-uninit](https://github.com/TritonVM/triton-vm/tree/asz/parallel-table-extension-option-maybe-uninit)

On my machine, running `prove_fib` with argument 500 (giving FRI domain length $2^{16}$) it generates these results:

| strategy | time (ms) |
|-------------|----------------|
| zeros+f() | 18.05 |
| set_len | 10.48 |
| maybe uninit | 15.47 |

Note that a benchmark more accurately measuring the target step is given by a recursive proof instance, which currently only works on Mjolnir.

Reply:
Also added 
 - [asz/parallel-table-extension-option-into-shape](https://github.com/TritonVM/triton-vm/tree/asz/parallel-table-extension-option-into-shape)

with timing
| strategy | time (ms) |
|-------|------|
| into shape | 18.3 |

This is despite the benchmark showing that "into shape" consistently outperforms the other options -- by an order of magnitude. So there is still a problem with the benchmarks that cause them to measure features not representative of practice.

Reply:
Benchmarking results on Mjolnir, prove fib 8000 FRI domain length $2^{20}$:

| strategy | time (ms) |
|---|---|
| zeros+f() | 219.08 |
| set_len | 47.34 |
| maybe uninit | 51.72 |
| into shape | 283.32 |

Some notes:
 - I also benchmarked the standalone benchmark `initialize_array` on Mjolnir and there the results point overwhelmingly in favor of "into shape" -- *up to ~1.5 M elements or 12 MB, after which `set_len` and `maybe_uninit` perform better.
 - The present benchmark, `prove_fib`, is more representative of practical deployment so that should be taken as higher prescriptive value.
 - It would be nice to fix benchmark `initialize_array` to be in line with practical deployment, but that is not a high priority.
 - In ordinal terms these timings are in line with what I observed on my own computer.
 - Strategy "set_len" is not really an option because its behavior is undefined.

Going forward I will set convenience function `fast_zeros_column_major` to use "maybe uninit".

Reply:
Addressed in 2812571226204fd2cd5e79df7cb53c176b83fadc. Resolving.